components/logins/src/db.rs

```diff
@@ -88,71 +91,73 @@ impl LoginDb {
-        db.set_pragma("key", encryption_key)?;
-        sqlcipher_3_compat(&db)?;
-        let salt = db.query_one::<String>("PRAGMA cipher_salt")?;
-        Ok(salt)
-    }
+        new_encryption_key: &str,
+        sqlcipher_path: impl AsRef<Path>,
+        sqlcipher_key: &str,
+        salt: Option<&str>,
+    ) -> Result<(Self, MigrationMetrics)> {
+        let path = path.as_ref();
+        let sqlcipher_path = sqlcipher_path.as_ref();
+
+        let metrics = if sqlcipher_path.exists() {
+            log::info!(
+                "Migrating sqlcipher DB: {} -> {}",
+                sqlcipher_path.display(),
+                path.display()
+            );
+            let result = migrate_sqlcipher_db_to_plaintext(
+                &sqlcipher_path,
+                &path,
+                sqlcipher_key,
+                new_encryption_key,
+                salt,
+            );
 
-    pub fn open_and_migrate_to_plaintext_header(
-        path: impl AsRef<Path>,
-        encryption_key: &str,
-        salt: &str,
-    ) -> Result<()> {
-        ensure_valid_salt(salt)?;
-        // Open the connection defensively without attempting to create a db if it doesn't exist.
-        let db = Connection::open_with_flags(path, OpenFlags::SQLITE_OPEN_READ_WRITE)?;
-        db.set_pragma("key", encryption_key)?;
-        sqlcipher_3_compat(&db)?;
-        db.set_pragma("cipher_salt", format!("x'{}'", salt))?;
-        // This tricks the `cipher_plaintext_header_size` command to work properly.
-        let user_version = db.query_one::<i64>("PRAGMA user_version")?;
-        // Remove the salt from the database header.
-        db.set_pragma("cipher_plaintext_header_size", 32)?;
-        // Flush the header changes.
-        db.set_pragma("user_version", user_version)?;
-        db.close().map_err(|(_conn, err)| err)?;
-        Ok(())
-    }
+            match result {
+                Err(e) => {
+                    log::error!("Error migrating sqlcipher DB: {}", e);
+                    // Delete both the old and new paths (if they exist)
+                    log::warn!("Re-creating database from scratch");
+                    if sqlcipher_path.exists() {
+                        std::fs::remove_file(sqlcipher_path)?;
+                    }
+                    if path.exists() {
+                        std::fs::remove_file(&path)?;
+                    }
+                    MigrationMetrics { ..Default::default()}
+                }
+                Ok(metrics) => {
+                    log::info!("Deleting old sqlcipher DB after migration");
+                    if sqlcipher_path.exists() {
+                        std::fs::remove_file(sqlcipher_path)?;
+                    }
+                    metrics
+                }
+            }
+        } else {
+            log::debug!("SQLCipher DB not found, skipping migration");
+            MigrationMetrics { ..Default::default()}
```

Comment:
I think `cargo fmt` is going to have something to say about this :)

Reply:
It complained, but maybe not enough.  I switched this line to `MigrationMetrics::default()`, along with a bunch of other `default()` calls.